File: 0437-path-sum-iii/0437-path-sum-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int targetSum) { 
        Map<Long, Integer> mapPrefixSum = new HashMap<>();
        mapPrefixSum.put(0L, 1);
        return prefixSum(root, targetSum, 0, mapPrefixSum);
    }

    private int prefixSum(TreeNode root, int targetSum, long currSum, Map<Long,Integer> mapPrefixSum) {
        if(root == null)
            return 0;
        currSum += root.val ;
        int count = 0;
        if(mapPrefixSum.containsKey(currSum-targetSum)){
            count += mapPrefixSum.get(currSum-targetSum);
        }
         
        mapPrefixSum.merge(currSum, 1, Integer::sum);
        int left = prefixSum(root.left,targetSum,currSum,mapPrefixSum) ;
        int right = prefixSum(root.right,targetSum,currSum,mapPrefixSum);
        
        mapPrefixSum.merge(currSum, -1, Integer::sum);

        return count+left+right;
    }

}

File: 0437-path-sum-iii/0437-path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        self.cnt=0
        def sm(no,ts):
            if no is None:
                return
            if no.val==ts:
                self.cnt+=1
            sm(no.left,ts-no.val)
            sm(no.right,ts-no.val)
        def dfs(node,tg):
            if node is None:
                return

            sm(node,tg)
            dfs(node.left,tg)
            dfs(node.right,tg)
        dfs(root,targetSum)
        return self.cnt